ldms/src/sampler/llnl/lustre_client_general.c

```diff
@@ -0,0 +1,276 @@
+ * Copyright (c) 2019 Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory. Written by
+ * Christopher J. Morrone <morrone2@llnl.gov>
+ * All rights reserved.
+ *
+ * This work was performed under the auspices of the U.S. Department of
+ * Energy by Lawrence Livermore National Laboratory under
+ * Contract DE-AC52-07NA27344.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the BSD-type
+ * license below:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *      Redistributions of source code must retain the above copyright
+ *      notice, this list of conditions and the following disclaimer.
+ *
+ *      Redistributions in binary form must reproduce the above
+ *      copyright notice, this list of conditions and the following
+ *      disclaimer in the documentation and/or other materials provided
+ *      with the distribution.
+ *
+ *      Neither the name of Sandia nor the names of any contributors may
+ *      be used to endorse or promote products derived from this software
+ *      without specific prior written permission.
+ *
+ *      Neither the name of Open Grid Computing nor the names of any
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ *      Modified source versions must be plainly marked as such, and
+ *      must not be misrepresented as being the original software.
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdio.h>
+#include <stdint.h>
+#include <dirent.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "ldms.h"
+#include "ldmsd.h"
+#include "ldms_jobid.h"
+#include "lustre_client.h"
+#include "lustre_client_general.h"
+
+static ldms_schema_t llite_general_schema;
+
+/* all of these will be prefixed with "stats." in the schema */
+static char *llite_stats_uint64_t_entries[] = {
```

Comment:
It would eliminate thousands of ldms_metric_by_name calls if you change ..._entries to
struct entry { const char *name, int idx; };
static struct entry llite_stats_uint64_t_entries[] = {
 { "dirty_pages_hits", -1},
 ...
};
and then at metric registration, set the idx field to the return code from ldms_schema_metric_add, and then at metric setting, just use the idx value.

Reply:
That doesn't quite solve the problem though.  Be aware that the list of values in the lustre stats file is not a static list of values.  I can't predict which will be there and which will not.  I might make an assumption of order, but I think it is best to avoid doing that as well.

With the proposed change,  I still need to do a linear walk on the array to find the index number each time I see it in the lustre stats file.  That seems no better than simply using ldms_metric_by_name().  

If we ever thought the linear walk was a real issue, then it would seem to me that implementing a hashed lookup under ldms_metric_by_name() would be the way to do it, rather than making each plugin implement the same thing.

Reply:
I'm not clear on why the list of metrics is not static. The *plugin reported* list must be static-- that's why we have a schema. You can define the convention for handling missing metrics as found in the lustre data (0, -1, etc; we generally favor 0 and/or setting a flag elsewhere in the data set if something is expected but missing. Alternatively, if the content for a given lustre installation may be any subset of the list you have in the code depending on the exact lustre configuration and version, you can define the schema by those metrics present when you start the sampler. If this is the case, append a checksum of the sorted metric names collectable to the user provided schema name before schema creation-- this will ensure there is never a schema abiguity. 
You can avoid linear searches using bsearch and defining your entries list array in sorted order and use -1 for the entries not seen by schema creation time.

Reply:
The metric contents are static, but the data from /proc is not.  In the metric I statically include all _possible_ values that might be seen in the "stats" file in /proc.  Lustre only starts printing out a stat once someone has actually _done_ that operation at least once.  This means that the various values only gradually start to appear as the mount is used.

Only including the metrics found at plugin start up time would be very, very wrong, because it is very likely that _none_ of the values will show anything at start up time.  And the the number of values in the stats file will almost certainly grow over time, so we should really just have all of them listed in the ldms metric set at start up time.

This kind of sucks for monitoring purposes.  We would probably prefer that lustre it just print out a line with a zero value instead of not printing the line at all.  In newer lustre proc files, such as "job_stats", they employ a new json format and they _do_ always include values of zero rather than printing nothing.  So they seem to have learned, but not updated old stats to work that way.

The long and short of it is that when I parse the lustre stats file, I need to look up the index number by the name I find in the stats file.  I can't make an simple numerical transformation from the line number in the stats file to the index number in the metric set.